esmvalcore/esgf/_download.py

```diff
@@ -0,0 +1,450 @@
+        else:
+            hasher = hashlib.new(checksum_type)
+
+        tmp_file = self._tmp_local_file(local_file)
+
+        logger.debug("Downloading %s to %s", url, tmp_file)
+        start_time = datetime.datetime.now()
+        response = requests.get(url,
+                                stream=True,
+                                timeout=TIMEOUT,
+                                cert=get_credentials())
+        response.raise_for_status()
+        with tmp_file.open("wb") as file:
+            for chunk in response.iter_content(chunk_size=None):
+                if hasher is not None:
+                    hasher.update(chunk)
+                file.write(chunk)
+
+        duration = datetime.datetime.now() - start_time
+
+        if hasher is None:
+            logger.warning(
+                "No checksum available, unable to check data"
+                " integrity for %s, ", url)
+        else:
+            local_checksum = hasher.hexdigest()
+            if local_checksum != checksum:
+                raise DownloadError(
+                    f"Wrong {checksum_type} checksum for file {tmp_file},"
+                    f" downloaded from {url}: expected {checksum}, but got"
+                    f" {local_checksum}. Try downloading the file again.")
+
+        shutil.move(tmp_file, local_file)
+        log_speed(url, self.size, duration.total_seconds())
+        logger.info("Downloaded %s (%.0f MB) in %s (%.1f MB/s) from %s",
+                    local_file, self.size / 2**20, duration,
+                    self.size / 2**20 / duration.total_seconds(),
+                    urlparse(url).hostname)
+
+
+def get_download_message(files):
+    """Create a log message describing what will be downloaded."""
+    megabyte = 2**20
+    gigabyte = 2**30
+    total_size = 0
+    lines = []
+    for file in files:
+        total_size += file.size
+        lines.append(f"{file.size / megabyte:.0f} MB" "\t" f"{file}")
+    if total_size:
+        lines.insert(0, "Will download the following files:")
+    lines.insert(0, f"Will download {total_size / gigabyte:.1f} GB")
+    return "\n".join(lines)
+
+
+def download(files, dest_folder, n_jobs=4):
+    """Download multiple ESGFFiles in parallel.
+
+    Arguments
+    ---------
+    files: list of :obj:`ESGFFile`
+        The files to download.
+    dest_folder: Path
+        The destination folder.
+    n_jobs: int
+        The number of files to download in parallel.
+
+    Raises
+    ------
+    DownloadError:
+        Raised if one or more files failed to download.
+    """
+    files = sorted(files)
+    logger.info(get_download_message(files))
+    logger.info("Downloading..")
+
+    def _download(file: ESGFFile):
+        """Download file to dest_folder."""
+        file.download(dest_folder)
+
+    total_size = 0
+    start_time = datetime.datetime.now()
+
+    errors = []
+    random.shuffle(files)
+    with concurrent.futures.ThreadPoolExecutor(max_workers=n_jobs) as executor:
```

Comment:
I'm not sure what the best number is, only experience can tell. However, I think that 4 is reasonable because even if you're downloading 4 files from the same host, you're probably not asking for too much data at the same time. In the future we could improve this, for example: download 2 files at the same time per host, so use (number of hosts * 2) threads.

Reply:
OKm sounds like a reasonable number then, here's an idea for superfast downloads http://eventlet.net/ - it could be that the bottleneck is the file writing anyway

Reply:
here's an example of a [web crawler](http://eventlet.net/doc/examples.html#producer-consumer-recursive-web-crawler)